[mlir][mpi] Lowering Mpi To LLVM

fschlimb opened the pull request:
As agreed by @AntonLydike this replaces #95524.

* The first set of patterns to convert the MPI dialect to LLVM.
* Further conversion pattern will be added in future PRs.

This adds the following on top of #95524 
* Support for Intel MPI by introducing MPIImplTraits to distinguish MPI implementations
* Works end-to-end in a downstream project using Intel MPI (going through Mesh dialect)
* Lowering MPI_Send/MPI_Recv
* Conversion will only be included if MPI is found

cc @mofeing @wsmoses @tobiasgrosser @hhkit @Groverkss @sjw36 @joker-eph @zero9178

AntonLydike committed:
add initial set of lowerings for MPI dialect

fschlimb committed:
lowering MPI_send and MPI_recv; some refactoring
fixing MPI_Send/Recv signature: added size
fixing MPIR ops.mlir tests
MPI: fixing enum names and merge conflicts
separate MPIImplTraits.h

fschlimb committed:
fixed finding MPI in cmake for MPIToLLVM
MPIImplTraits for OpenMPI

fschlimb commented:
> > Support for Intel MPI by introducing MPIImplTraits to distinguish MPI implementations
> 
> Next MPI standard stabilizes the ABI, but I guess we still need to support earlier vendor-ABIs of earlier MPI implementations right?
> 
> cc @JBlaschke @hhkit

Yes, we can move to that once that is broadly available.

fschlimb committed:
provide empty register/populate functions even when MPIToLLVM is disabled

fschlimb committed:
fallbacks from MPICH if no mpi.h was found

AntonLydike commented on mlir/test/Conversion/MPIToLLVM/ops.mlir:
```diff
@@ -0,0 +1,84 @@
+    // CHECK-NEXT: [[v25:%.*]] = llvm.getelementptr [[v23]][[[v24]]] : (!llvm.ptr, i64) -> !llvm.ptr, f32
+    // CHECK-NEXT: [[v26:%.*]] = llvm.extractvalue [[v5]][3, 0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
+    // CHECK-NEXT: [[v27:%.*]] = llvm.trunc [[v26]] : i64 to i32
+    // CHECK-NEXT: [[v28:%.*]] = llvm.mlir.
+    // CHECK-NEXT: [[v29:%.*]] = llvm.mlir.
```
these appear to be missing something?

fschlimb replied:
Yes, the constants used in there depend on the MPI implementation. This makes sure the test passes for any MPI.

AntonLydike reviewed (approved):
:heart: Really cool work! Thanks @fschlimb! Super excited to see this progress!

Dinistro commented on mlir/include/mlir/Conversion/MPIToLLVM/MPIToLLVM.h:
```diff
@@ -0,0 +1,31 @@
+//
```
Nit: Broken banner

Dinistro commented on mlir/include/mlir/Conversion/MPIToLLVM/MPIToLLVM.h:
```diff
@@ -0,0 +1,31 @@
+class LLVMTypeConverter;
+class RewritePatternSet;
+
+#define GEN_PASS_DECL_MPITOLLVMCONVERSIONPASS
+#include "mlir/Conversion/Passes.h.inc"
```
I don't understand. Why is this required when there is no actual pass that backs this. This is an interface based pass extension, no?

fschlimb replied:
Left-over. Can be removed.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,332 @@
+
+using namespace mlir;
+
+// TODO: this was copied from GPUOpsLowering.cpp:288
+// is this okay, or should this be moved to some common file?
```
This should be moved into Conversion/LLVMCommon, I presume

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/mpi_fallback.h:
This might be problematic to include here given the different licence. I'm not sure how this should be handled, tbh.

fschlimb replied:
The MPICH license is compatible because it is fully permissive. Should anyone see an issue with this, I can write it myself and make up my own values. We'd lose compatibility with MPICH/Intel MPI, though.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIImplTraits.h:
```diff
@@ -0,0 +1,157 @@
+// OpenMPI
+#elif defined(OPEN_MPI) && OPEN_MPI == 1
+
+// TODO: this is pretty close to getOrDefineFunction, can probably be factored
```
Nit: Please do

fschlimb replied:
done

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIImplTraits.h:
What's the benefit of factoring this into a header file?

fschlimb replied:
It's indeed mostly a style issue. I like outlining details that are mostly in the way and confusing without contributing the actual mechanics. This looks much cleaner to me.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIImplTraits.h:
```diff
@@ -0,0 +1,157 @@
+#include "mlir/Transforms/DialectConversion.h"
+
+// skip if no MPI C header was found
+#ifdef FOUND_MPI_C_HEADER
+#include <mpi.h>
+#else // not FOUND_MPI_C_HEADER
+#include "mpi_fallback.h"
+#endif // FOUND_MPI_C_HEADER
```
I'm not sure if this is acceptable, tbh. This should probably be a configuration option of the pass, not something taking from header files. This is especially problematic when building on one machine and then distributing the produced binaries to some other system which might now have these headers.

fschlimb replied:
Are you suggesting to make this "runtime" options, e.g. chose the right values/configuration when the compiler is running?
I agree, that would of course be ideal. I suspect that that it would overcomplicate things in the current state of affairs. At this point, the MPI support is very nascent, and MLIR is rarely (if ever) distributed in binary form. 
We can always improve in future PRs, in particular if the new ABI turns out to be insufficient or is taking too long to materialize.

fschlimb replied:
I am not aware of a tangible plan. Probably mostly because no concrete suggestion was made yet. 
I'd be happy to work with you on that, but I suggest doing this in a separate PR.

Notice: Even though not ideal, the current default/fallback approach works for all the MPICH compatible MPI implementations (https://www.mpich.org/abi/).

Dinistro reviewed (changes requested):
Very nice work and sorry for the late review.

Unfortunately, there are some issues with the dependencies on MPI headers and packages. I highly doubt that this is an acceptable thing to depend on in the build system. Note that test might break when someone has another header around that depends different values for the used constants.

I suggest to restructure the pass to take all the necessary values as configuration parameters. In the worst case, this could even be a json file, if the list of options is growing. I'm not really sure who could consult on this.

fschlimb commented:
> Very nice work and sorry for the late review.
> 
> Unfortunately, there are some issues with the dependencies on MPI headers and packages. I highly doubt that this is an acceptable thing to depend on in the build system. Note that test might break when someone has another header around that depends different values for the used constants.
> 
> I suggest to restructure the pass to take all the necessary values as configuration parameters. In the worst case, this could even be a json file, if the list of options is growing. I'm not really sure who could consult on this.

Yes, there is no straight-forward solution for this right now. We could depend on yet another (external) library which unifies the interfaces. To me that's not really tempting for getting a few defines.

Why would configuration parameters change anything about the issue you are mentioning? Whatever was configured can or cannot conflict later, no matter what. Notice that later is not when the compiler is built, but when the compiler is executed. If you are concerned about other passes/dialects using a different MPI in the same build, we could move finding the header-file to an upper level cmake file. Frankly, I don't see this as a problem.

As mentioned above, as soon as the new MPI standard defines a stable ABI this problem is gone anyway. I suggest keeping this like this for the time being.

fschlimb commented on mlir/lib/Conversion/MPIToLLVM/CMakeLists.txt:
What do you mean by "compatible"?

fschlimb replied:
Yes.

fschlimb commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,332 @@
+  matchAndRewrite(mpi::InitOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // get loc
+    auto loc = op.getLoc();
```
guidance is: "Don’t “almost always” use auto, but do use auto with initializers like cast<Foo>(...) or other places where the type is already obvious from the context."
I guess it lies in the eye of the beholder, but to me the type is clearly obvious from the context.

fschlimb replied:
Anyway, I changed as you suggested.

fschlimb commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,332 @@
+  matchAndRewrite(mpi::CommRankOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // get some helper vars
+    auto loc = op.getLoc();
+    auto context = rewriter.getContext();
+    auto i32 = rewriter.getI32Type();
```
See above.

fschlimb commented on mlir/test/Conversion/MPIToLLVM/ops.mlir:
```diff
@@ -0,0 +1,84 @@
+    // CHECK: llvm.call @MPI_Finalize() : () -> i32
+    %3 = mpi.finalize : !mpi.retval
+
+    %4 = mpi.retval_check %retval = <MPI_SUCCESS> : i1
+
+    %5 = mpi.error_class %0 : !mpi.retval
```
We currently do not handle these. Can add a check nevertheless.

fschlimb replied:
removed.

fschlimb committed:
formatting and cleanup (review comments)

fschlimb commented:
Thanks a lot @Dinistro for your valuable comments and questions.
I applied most of your suggestions and replied to others.

Dinistro reviewed (commented):
> Why would configuration parameters change anything about the issue you are mentioning? Whatever was configured can or cannot conflict later, no matter what. Notice that later is not when the compiler is built, but when the compiler is executed. If you are concerned about other passes/dialects using a different MPI in the same build, we could move finding the header-file to an upper level cmake file. Frankly, I don't see this as a problem.

I was referring to lowering pass options that can be passed into the compiler, i.e., compiling to a specific MPI implementation/draft instead of tying the supported MPI implementation to the configuration time of MLIR itself.

In any case, I don't feel confident to judge if this is an acceptable workaround or not. Thus, I'm explicitly requesting someone from core to weight in. CC @joker-eph @matthias-springer @ftynse

fschlimb committed:
inlining MPIImplTraits and fallback into cpp file, minor mods addressing review comments

fschlimb commented on mlir/test/Conversion/MPIToLLVM/ops.mlir:
```diff
@@ -0,0 +1,83 @@
+  // CHECK: llvm.call @MPI_Finalize() : () -> i32
+  %3 = mpi.finalize : !mpi.retval
+
+  // CHECK: mpi.retval_check
```
Removed the ops, they are unnecessary.

fschlimb commented on mlir/test/Conversion/MPIToLLVM/ops.mlir:
```diff
@@ -0,0 +1,83 @@
+
+  // CHECK: mpi.retval_check
+  %4 = mpi.retval_check %retval = <MPI_SUCCESS> : i1
+  // CEHCK: mpi.error_class
```
Removed.

fschlimb commented on mlir/lib/Conversion/MPIToLLVM/mpi_fallback.h:
Inlined into cpp.

fschlimb commented:
> Many places in MLIR adapted a very clear stance on this. I'm aware that this rule is quite flexible, which makes it a bad one.

Whatever. I changed as you suggested.

fschlimb committed:
rm unused line

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

fschlimb commented on mlir/test/Conversion/MPIToLLVM/ops.mlir:
```diff
@@ -0,0 +1,84 @@
+// RUN: mlir-opt -convert-to-llvm %s | FileCheck %s
+
+module {
```
removed.

tobiasgrosser commented:
> > Why would configuration parameters change anything about the issue you are mentioning? Whatever was configured can or cannot conflict later, no matter what. Notice that later is not when the compiler is built, but when the compiler is executed. If you are concerned about other passes/dialects using a different MPI in the same build, we could move finding the header-file to an upper level cmake file. Frankly, I don't see this as a problem.
> 
> I was referring to lowering pass options that can be passed into the compiler, i.e., compiling to a specific MPI implementation/draft instead of tying the supported MPI implementation to the configuration time of MLIR itself.
> 
> In any case, I don't feel confident to judge if this is an acceptable workaround or not. Thus, I'm explicitly requesting someone from core to weight in. CC @joker-eph @matthias-springer @ftynse

Instead of linking in the MPI header file that is decided on by the pre-processor, could we make the MPI library a run-time choice? Instead of using a header file, I suggest we hardcode the ABI based on our knowledge. We can use a little python script to extract the ABI we need, but requiring the mpi.h header might not be a good idea. In the end, 90% of this PR already hardcodes the MPI API without relying on the MPI headers. I feel the cost of adding some ABI knowledge is small and the benefit of being independent of headers and in-line with MLIR-style are notable.

fschlimb commented:
> Instead of linking in the MPI header file that is decided on by the pre-processor, could we make the MPI library a run-time choice? Instead of using a header file, I suggest we hardcode the ABI based on our knowledge. We can use a little python script to extract the ABI we need, but requiring the mpi.h header might not be a good idea. In the end, 90% of this PR already hardcodes the MPI API without relying on the MPI headers. I feel the cost of adding some ABI knowledge is small and the benefit of being independent of headers and in-line with MLIR-style are notable.

Yes, as the "fallback" shows, currently there is not much we need to know. We can do the same for OpenMPI. Of course this raises the same license questions as for MPICH. I am convinced both licenses are sufficiently permissive.

Adding "runtime" option is of course more work. Anyone willing to help here? Or can we do this in a separate PR?

fschlimb commented:
> > Any objections to addressing this in a separate PR?
> 
> IIUC, the current lowering pass changes behavior depending on the MPI version that was around during build time. If that is really the case, then the test should not work properly depending on the build setting, no?

The tests are written in a way, that the differences between OpenMPI and MPICH* are hidden. This is mostly about different constant values. @AntonLydike also found places where details of checked operations are missing (and only the operation name itself gets verified).

fschlimb commented:
I thought a bit about the runtime option. I see only two ways of doing this in a reasonable way:

1. I found no way of adding  custom option to `--convert-to-llvm`. So if we want this to become an option to the pass, we need a separate pass (and not populate the generic  `--convert-to-llvm` pass). There seems to be some consensus that the current approach is preferred (and In #95524 @joker-eph explicitly asked for it). Any hint on how to add an option to `--convert-to-llvm` would be appreciated.
2. Use DLTI to annotate the IR. This could be something like the below. I prefer this solution, it is more flexible (e.g. one could write an extra pass wich inserts the DLTI according to an option to the pass) and keeps the integration into `--convert-to-llvm`.

```mlir
module attributes { mpi.dlti = #dlti.map<"MPI:Implemention" = "Intel"> } {
  %0 = mpi.....
}
```

Any thoughts on this?

fschlimb committed:
remove dependency on mpi.h; TODO: runtime dispatch

tobiasgrosser commented:
Nice!

fschlimb committed:
clang-format

fschlimb committed:
MPI implementation selection at runtime

fschlimb commented:
@Dinistro @tobiasgrosser I just (force-)pushed a solution which selects the MPI implementation at runtime. As suggested above, it uses DLTI. Hope this addresses your concerns.

rolfmorel commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -184,29 +185,53 @@ struct OMPIImplTraits {
+    if (strAttr && strAttr.getValue() == "OpenMPI") {
+      return OMPI;
+    }
+    return MPICH;
```
It seems to me that if we get to this point the user specified yet another OpenMPI implementation. Hard for me to judge whether it is okay to proceed as MPICH, though a warning that you are ignoring whatever was specified could considerably help debugging (e.g. someone typed "openmpi" as the key name or "OpenMP", etc).

rolfmorel replied:
On that note: you probably want to check if the value string is "MPICH" as in that case you _don't_ want to output a warning. To state the obvious. 😅

fschlimb replied:
ok, emitting a waring if specified attribute is unexpected.

rolfmorel commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -184,29 +185,53 @@ struct OMPIImplTraits {
                                   const mlir::Location loc,
                                   mlir::ConversionPatternRewriter &rewriter) {
-    // TODO: dispatch based on the MPI implementation
+    if (MPIImplTraits::getMPIImpl(moduleOp) == OMPI) {
```
Depending on how often `getCommWorld` and `getStatusIgnore` get invoked, I would recommend "pulling up" the call to `getMPIImpl` as far as possible, e.g. if feasible to `runOnOperation` / outermost function of your code which is first handed the ModuleOp. This is because there are instances where calling `dlti::query` could be expensive.

Having said that, these particular calls to `dlti::query` are really the cheapest possible invocation of it (no traversal of ancestors - no nested attributes), so I very much doubt it would be problematic unless `getCommWorld` and `getStatusIgnore` get invoked often.

fschlimb replied:
I presume performance optimizations are premature at this stage. If it becomes an issue, we can look at it.

rolfmorel replied:
I am fine with that argument :+1:

fschlimb committed:
warning about unkonw MPi implementation

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+// Implementation details for MPICH ABI compatible MPI implementations
+//===----------------------------------------------------------------------===//
+struct MPICHImplTraits {
+  static const int MPI_FLOAT = 0x4c00040a;
```
```suggestion
  static constexpr int MPI_FLOAT = 0x4c00040a;
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// When lowering the mpi dialect to functions calls certain details
+// differ between various MPI implementations. This class will provide
+// these in a generic way, depending on the MPI implementation that got
+// selected by the DLTI attribute on the module.
+//===----------------------------------------------------------------------===//
```
```suggestion
/// When lowering the mpi dialect to functions calls certain details
/// differ between various MPI implementations. This class will provide
/// these in a generic way, depending on the MPI implementation that got
/// selected by the DLTI attribute on the module.
```
This should be turned into a normal docstring for the struct.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+  // Default to MPICH (and ABI compatible).
+  static MPIImpl getMPIImpl(mlir::ModuleOp &moduleOp) {
+    auto attr = dlti::query(*&moduleOp, {"MPI:Implementation"}, true);
+    if (failed(attr)) {
+      return MPICH;
+    }
```
```suggestion
    if (failed(attr))
      return MPICH;
```
Nit: The style guide requests that single statement blocks omit the braces.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+struct MPIImplTraits {
+  enum MPIImpl { MPICH, OMPI };
+
+  // Get the MPI implementation from a DLTI attribute on the module.
+  // Default to MPICH (and ABI compatible).
```
```suggestion
  /// Gets the MPI implementation from a DLTI attribute on the module.
  /// Defaults to MPICH (and ABI compatible).
```
Nit: In general, comments on classes and members should use `///` for doxygen to pick them up.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+      return OMPI;
+    }
+    if (!strAttr || strAttr.getValue() != "MPICH") {
+      moduleOp.emitWarning("Unknown \"MPI:Implementation\" specified in DLTI, "
+                           "defaulting to MPICH");
```
```suggestion
      moduleOp.emitWarning() << "Unknown \"MPI:Implementation\" value in DLTI (" << strAttr.getValue() << "), "
                           "defaulting to MPICH");
```
Nit: Mention the unexpected value.

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+    return MPICH;
+  }
+
+  // get/create MPI_COMM_WORLD as a mlir::Value
```
```suggestion
  /// Gets or creates MPI_COMM_WORLD as a mlir::Value.
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,521 @@
+  LogicalResult
+  matchAndRewrite(mpi::InitOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // get loc
```
```suggestion
```
Nit: This is not helping much, I think

Dinistro reviewed (commented):
Very nice, thanks for changing this to use the DLTI. I mainly added nit comments, related to things from the style guide: https://mlir.llvm.org/getting_started/DeveloperGuide/

Considering the new structure, it might be more elegant to rely on virtual dispatch to implement the different lowering strategies. This use case very nicely fits the strategy pattern, I believe.
This would ensure that you only need to check the DLTI once and can thereafter rely on the virtual dispatch to delegate to the proper implementation.

fschlimb committed:
Apply suggestions from code review

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

fschlimb commented:
> Very nice, thanks for changing this to use the DLTI. I mainly added nit comments, related to things from the style guide: https://mlir.llvm.org/getting_started/DeveloperGuide/
> 
> Considering the new structure, it might be more elegant to rely on virtual dispatch to implement the different lowering strategies. This use case very nicely fits the strategy pattern, I believe. This would ensure that you only need to check the DLTI once and can thereafter rely on the virtual dispatch to delegate to the proper implementation.

Yes, I have been thinking about this as well. Virtual dispatch requires an object. The easy lifetime of that object is each `matchAndRewrite`. What we ideally want is one object per module. Without a proper pass, I don't know how to do this.

Yes, the current solution is not perfect. We can go on like this forever, or we declare this as good enough as a first step and together work on continuous improvements in future PRs. I vote for the latter.

fschlimb committed:
fixing formatting and typo

fschlimb committed:
using virtual dispatch for MPIImplTraits; cleanup

fschlimb commented:
Thanks a lot for your review and suggestions, @Dinistro !
I have converted the Traits classes to use virtual functions for dispatch. Each conversion pattern invocation creates its own trait object (once). This is not idea in terms of performance, but much better than before. 
I hope the latest commit addresses all your remaining concerns and you are ok with merging.

fschlimb committed:
without including mpi.h leading _s are not needed

Dinistro commented on mlir/test/Conversion/MPIToLLVM/ops.mlir:
```diff
@@ -0,0 +1,165 @@
+// RUN: mlir-opt -split-input-file -convert-to-llvm %s | FileCheck %s
+
+// COM: Test MPICh ABI
```
```suggestion
// COM: Test MPICH ABI
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+//
+//===----------------------------------------------------------------------===//
+
+//
+// Copyright (C) by Argonne National Laboratory
+//    See COPYRIGHT in top-level directory
+//    of MPICH source repository.
+//
```
I guess this is no longer needed?

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+  /// on the given module.
+  static std::unique_ptr<MPIImplTraits> get(ModuleOp &moduleOp);
+
+  MPIImplTraits(ModuleOp &moduleOp) : moduleOp(moduleOp) {}
```
```suggestion
  explicit MPIImplTraits(ModuleOp &moduleOp) : moduleOp(moduleOp) {}
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+  /// Get the MPI_STATUS_IGNORE value (typically a pointer type).
+  virtual intptr_t getStatusIgnore() = 0;
+
+  /// get/create MPI datatype as a Value which corresponds to the given
+  /// Type
```
```suggestion
  /// Gets or creates an MPI datatype as a value which corresponds to the given
  /// type.
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+
+  Value getCommWorld(const Location loc,
+                     ConversionPatternRewriter &rewriter) override {
+    static const int MPI_COMM_WORLD = 0x44000000;
```
```suggestion
    static constexpr int MPI_COMM_WORLD = 0x44000000;
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+    StringRef name = "ompi_mpi_comm_world";
+
+    // make sure global op definition exists
+    (void)getOrDefineExternalStruct(loc, rewriter, name, commStructT);
```
```suggestion
    getOrDefineExternalStruct(loc, rewriter, name, commStructT);
```
Nit: Should not be necessary, as the function is not marked with `nodiscard`

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+    auto commStructT =
+        LLVM::LLVMStructType::getOpaque("ompi_predefined_datatype_t", context);
+    // make sure global op definition exists
+    (void)getOrDefineExternalStruct(loc, rewriter, mtype, commStructT);
```
```suggestion
    getOrDefineExternalStruct(loc, rewriter, mtype, commStructT);
```

Dinistro commented on mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:
```diff
@@ -0,0 +1,501 @@
+  ModuleOp &moduleOp;
+
+public:
+  /// Instantiate a new MPIImplTraits object according to the DLTI attribute
+  /// on the given module.
```
Maybe clarify that the fallback is MPICH

fschlimb replied:
done

Dinistro reviewed (approved):
Very nice work, thanks for addressing all the comments. I added a few more nits but this is essentially ready to go 🙂

fschlimb committed:
Apply suggestions from code review

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

fschlimb committed:
clarified MPICH default

fschlimb commented:
Excellent. Thanks again @AntonLydike, @Dinistro, @tobiasgrosser and @mofeing for your comments!

kazutakahirata commented:
With this PR, I am getting:

```
/usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:93:2: error: delete called on '(anonymous namespace)::MPIImplTraits' that is abstract but has non-virtual destructor [-Werror,-Wdelete-abstract-non-virtual-dtor]
        delete __ptr;
        ^
/usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:398:4: note: in instantiation of member function 'std::default_delete<(anonymous namespace)::MPIImplTraits>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:313:22: note: in instantiation of member function 'std::unique_ptr<(anonymous namespace)::MPIImplTraits>::~unique_ptr' requested here
    auto mpiTraits = MPIImplTraits::get(moduleOp);
```

Is there any way you could take a look?  My cmake configuration is `-DLLVM_ENABLE_WERROR=On` with clang being the host compiler.  Thanks in advance!

kazutakahirata commented:
@fschlimb I've fixed the warnings with 386a45c6bd6358128f1e8e0819bacc92767f8db6.  Thanks!

fschlimb commented:
> @fschlimb I've fixed the warnings with [386a45c](https://github.com/llvm/llvm-project/commit/386a45c6bd6358128f1e8e0819bacc92767f8db6). Thanks!

@kazutakahirata Thanks so much!
